pkg/typesystem/typesystem.go

```diff
@@ -411,33 +414,37 @@ func (t *TypeSystem) IsDirectlyRelated(target *openfgav1.RelationReference, sour
 	return false, nil
 }
 
-// TTUResolvesExclusivelyToDirectlyAssignable returns whether the computedRelation of object's tupleRelation is directly assignable.
-func (t *TypeSystem) TTUResolvesExclusivelyToDirectlyAssignable(objectType, tuplesetRelation, computedRelation string) (bool, error) {
-	tuplesetRelationTypes, directlyAssignable, err := t.resolvesTypeRelationToDirectlyAssignable(objectType, tuplesetRelation)
-	if err != nil {
-		return false, err
-	}
-	if !directlyAssignable {
-		return false, nil
-	}
-	var relationUndefinedError *RelationUndefinedError
-	for _, tuplesetRelationType := range tuplesetRelationTypes {
-		_, childDirectlyAssignable, err := t.resolvesTypeRelationToDirectlyAssignable(tuplesetRelationType, computedRelation)
-		if err != nil {
-			// in the case of errors due to relation undefined, we can ignore the error because it is possible
-			// that some parents do not have the relation defined.
-			if errors.As(err, &relationUndefinedError) {
+// TTUCanFastPath returns whether object's tupleRelation's rewrite can support the fast path optimization.
+func (t *TypeSystem) TTUCanFastPath(objectType, computedRelation, userType string) bool {
+	tuplesetRelation := t.relations[objectType][computedRelation].GetRewrite().GetTupleToUserset().GetTupleset().GetRelation()
+
+	computedUsersetRelation := t.relations[objectType][computedRelation].GetRewrite().GetTupleToUserset().GetComputedUserset().GetRelation()
+	ttuParentTypes := t.relations[objectType][tuplesetRelation].GetTypeInfo().GetDirectlyRelatedUserTypes()
+
+	if len(ttuParentTypes) > 1 {
+		// For TTU with multiple assignable types, need to verify that each type has the computed relation and is eligible for fast-path
+		for _, parentType := range ttuParentTypes {
+			_, relationExists := t.relations[parentType.GetType()][computedUsersetRelation]
+			if !relationExists {
```

Comment:
This swath of code is necessary for detecting cases where there are multiple possible TTU type restrictions defined where one or more of the tupleset relations are ineligible for fast-path. Example:

```
model
	schema 1.1
type user
type folder
	relations
	define owner: [user]
	define viewer: [user] or owner # TTU slow path because union
type document
	relations
	define can_read: viewer from parent
	define parent: [document, folder]
	define viewer: [user] # TTU fast path
```

The TTU relation `document#parent` directs to two relations, `document#viewer` and `folder#viewer`. The former is eligible for fast-path but the latter is not because of the union. Because of this, the entire `document#parent` relation is ineligible for fast-path resolution.